crates/shared/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -86,166 +100,365 @@ impl TryFrom<PoolData> for PoolInfo {
+    async fn update_missing_pools(&self) -> Result<()> {
+        let (missing_pools, block_number) = {
+            let checkpoint = self.pools_checkpoint.lock().unwrap();
+            (checkpoint.missing_pools.clone(), checkpoint.block_number)
+        };
+
+        let pool_ids = missing_pools.into_iter().collect::<Vec<_>>();
+        let pools = self
+            .graph_api
+            .get_pools_with_ticks_by_ids(&pool_ids, block_number)
+            .await?;
+
+        let mut checkpoint = self.pools_checkpoint.lock().unwrap();
+        for pool in pools {
+            checkpoint.missing_pools.remove(&pool.id);
+            checkpoint.pools.insert(pool.id, pool);
+        }
+        Ok(())
+    }
+}
+
+pub struct UniswapV3PoolFetcher {
+    /// Pools state on a specific block number in history considered reorg safe
+    checkpoint: PoolsCheckpointHandler,
+    /// Recent events used on top of pools_checkpoint to get the `latest_block` pools state.
+    events: tokio::sync::Mutex<EventHandler<Web3, UniswapV3PoolEventFetcher, RecentEventsCache>>,
+}
+
+impl UniswapV3PoolFetcher {
+    pub async fn new(chain_id: u64, client: Client, web3: Web3) -> Result<Self> {
+        let checkpoint = PoolsCheckpointHandler::new(chain_id, client).await?;
+
+        let init_block = checkpoint.pools_checkpoint.lock().unwrap().block_number;
+        let init_block = web3
+            .eth()
+            .block(BlockNumber::Number(init_block.into()).into())
+            .await?
+            .context("missing block for fetched block number")?;
+        let init_block = (
+            init_block
+                .number
+                .context("missing fetched block number")?
+                .as_u64(),
+            init_block.hash.context("missing fetched block hash")?,
+        );
+
+        let events = tokio::sync::Mutex::new(EventHandler::new(
+            web3.clone(),
+            UniswapV3PoolEventFetcher(web3.clone()),
+            RecentEventsCache::default(),
+            Some(init_block),
+        ));
+
+        Ok(Self { checkpoint, events })
+    }
+
+    /// Moves the checkpoint to the block `latest_block - MAX_REORG_BLOCK_COUNT`
+    async fn move_checkpoint_to_future(&self) -> Result<()> {
+        let last_event_block = self.events.lock().await.store().last_event_block().await?;
+        let old_checkpoint_block = self
+            .checkpoint
+            .pools_checkpoint
+            .lock()
+            .unwrap()
+            .block_number;
+        let new_checkpoint_block = std::cmp::max(
+            last_event_block.saturating_sub(MAX_REORG_BLOCK_COUNT),
+            old_checkpoint_block,
+        );
+
+        if new_checkpoint_block > old_checkpoint_block {
+            let block_range =
+                RangeInclusive::try_new(old_checkpoint_block + 1, new_checkpoint_block)?;
+            let events = self.events.lock().await.store().get_events(block_range);
```

Comment:
Do we need to remove old events from the event storage? Shouldn't we cleanup everything older than `MAX_REORG_BLOCK_COUNT` once we are done using it?

Reply:
I am not sure about this. My assumption was we never cleanup since we have some `EventHandler` instances that use db as an event store (like `EventUpdater` for example where the settlement contract swaps are saved)

Reply:
But aren't these events held in memory and not in the database? I think if they are in

Reply:
Yes, I just meant that if we want different clean up policies for each `EventHandler` instance then I'll need to add an option for that in the `EventHandler` constructor.

Reply:
It could be added as a separate PR.

Reply:
This is very specific to Uniswap V3 though - for Balancer V2 we only keep the equivalent of "checkpoint" anyway, so there is no `Vec<>` of events to cleanup.

My point is that, when do a checkpoint, the old events from before the checkpoint aren't valuable anymore and can just be cleaned up. I imagine it can be done directly from the `move_checkpoint_to_future` function right?

Reply:
Actually yes. Added.